src/EiffelEvents.Net/Validation/ValidationHelper.cs

```diff
@@ -0,0 +1,80 @@
+﻿//    Copyright (c) 2021 ICT Cube, doWhile, and Eiffel Community collaborators.
+// 
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+// 
+//        http://www.apache.org/licenses/LICENSE-2.0
+// 
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using EiffelEvents.Net.Events.Core;
+using EiffelEvents.Net.Exceptions;
+using FluentResults;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Schema;
+
+namespace EiffelEvents.Net.Validation
+{
+    public static class ValidationHelper
+    {
+        /// <summary>
+        /// Validate Event instance Version against event type version
+        /// </summary>
+        /// <param name="eiffelEvent">Instance to be checked</param>
+        /// <typeparam name="T">Type to check against</typeparam>
+        /// <exception cref="InvalidEiffelEventException">Thrown if the versions are not equal</exception>
+        public static void ValidateEventVersion<T>(T eiffelEvent) where T : IEiffelEvent
+        {
+            var typeObj = (T)Activator.CreateInstance(typeof(T));
+            if (eiffelEvent.GetVersion() != typeObj?.GetVersion())
+                throw new InvalidEiffelEventException(typeof(T).Name, eiffelEvent.GetVersion());
+        }
+
+        public static Result ValidateEventSchema<T>(string eiffelEventJson)
+            where T : IEiffelEvent
+        {
+            // load schema
+            var eventName = typeof(T).Name;
+            var edition = typeof(T).Namespace?.Split('.').Last().Replace('_', '-');
```

Comment:
This code is dangerous and may break when refactoring. Can it be solved some other way?

It is unlikely to ever cause any issues though, so I'd be fine with leaving it if it is not easy to solve.

Reply:
I guess another way of solving it would be to access the Version string from the type (or the JSON) and use that to find the schema.

Then it would not be connected to any specific edition, and would instead validate against what the event claims to be.

Reply:
Unfortunately, we are tightly coupled with the edition because of the deserialization step. For more clarification on removing the directory structure for editions we can face this scenario:
- Subscribe to an event from `Lyon-Edition`.
- An event JSON received.
- We looked up its version `3.0.0`, which means `Paris-Edition`, from JSON and the matched event schema was found based on that version.
- The event was validated successfully.
- Library will try to deserialize the JSON to the corresponding Eiffel C# type passed to the `Subscribe<namespace.T>()` method. If the type used in the `Subscribe` method was from the `Lyon-Edition` namespace, deserialization will fail.

Moreover, the method `ValidateEventSchema` is covered by tests, e.g. `ValidateSchema_Valid_Success`, and that test will fail if we changed the namespace.

Reply:
I'm not saying we need to de-serialize events into the wrong type, I'm saying that when we already have a JSON-format event, which is the case for this method, we could get the event type and version from the JSON data and use that to validate the event.

I think that we in this method should validate "does the schema of the outgoing event match its own stated type and version", not "does the incoming event have the right event type and version".

I may have misunderstood your commend though, as I thought we validated outgoing events, not incoming events.

Reply:
Was this resolved? It doesn't seem resolved to me, as the code is unchanged. I thought we would re-arrange to get the event type and version from the JSON, and use that to pick the right schema?

Reply:
Technically, it can be feasible, but it may have a little bit of bad performance to access a serialized event and get these pieces of information from it.